node/core/provisioner/src/onchain_disputes.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright 2017-2022 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg(feature = "improved-onchain-disputes-import")]
+
+use futures::channel::oneshot;
+use polkadot_node_subsystem::{
+	errors::RuntimeApiError,
+	messages::{RuntimeApiMessage, RuntimeApiRequest},
+	SubsystemSender,
+};
+use polkadot_primitives::v2::{CandidateHash, Hash, SessionIndex};
+use std::collections::HashSet;
+
+pub enum GetOnchainDisputesErr {
+	Channel,
+	Execution,
+	NotSupported,
+}
+
+/// Gets the on-chain disputes at a given block number and returns them as a `HashSet` so that searching in them is cheap.
+pub async fn get_onchain_disputes(
+	sender: &mut impl SubsystemSender,
+	relay_parent: Hash,
+) -> Result<HashSet<(SessionIndex, CandidateHash)>, GetOnchainDisputesErr> {
+	gum::trace!("Fetching on-chain disputes", ?relay_parent);
```

Comment:
please add a target for each log, so we can enable it only for one subsystem

Reply:
Done